src/main/java/org/truffleruby/core/rubinius/IONodes.java

```diff
@@ -805,4 +753,32 @@ public DynamicObject setLastLine(VirtualFrame frame, DynamicObject matchData) {
         }
     }
 
+    @Primitive(name = "io_get_thread_buffer", needsSelf = false)
+    public static abstract class GetThreadBufferNode extends PrimitiveArrayArgumentsNode {
+
+        private static final ThreadLocal<Pointer> THREAD_BUFFER_THREAD_LOCAL = ThreadLocal.withInitial(() -> Pointer.NULL);
+
+        @Specialization
+        public DynamicObject getThreadBuffer(long size,
+                                             @Cached("create()") AllocateObjectNode allocateObjectNode) {
+            return allocateObjectNode.allocate(getContext().getCoreLibrary().getRubiniusFFIPointerClass(), getBuffer(size));
+        }
+
+        @TruffleBoundary
+        private Pointer getBuffer(long size) {
+            final Pointer buffer = THREAD_BUFFER_THREAD_LOCAL.get();
+            if (buffer.size() >= size) {
+                return buffer;
+            } else {
+                if (!buffer.isNull()) {
+                    buffer.free();
+                }
+                final Pointer newBuffer = Pointer.malloc(Math.max(size * 2, 1024));
```

Comment:
This logic and the check above could be simplified if the initial value of the ThreadLocal is a pointer of 1024 bytes.

Reply:
I thought that then we'd have to possibly allocate the buffer twice on first use though.